### ISSUE TITLE
How to be defensive and avoid crashes on React Native?

### ISSUE THREAD

**Issue by thatjuan** [realm/realm-js#887]
I'm working on a React Native 0.40 app that uses Realm JS 1.0.0. I am experiencing crashes when reading/writing and some odd behaviors. For example, one of the crashes only happens when I add a second record to the table. 

When it crashes, the output prints:

    <Error>: MY_APP(18085,0x7000075ce000) malloc: *** mach_vm_map(size=18446744073345937408) failed (error code=3)
    	*** error: can't allocate region
    	*** set a breakpoint in malloc_error_break to debug

(this is on the iOS simulator)

I've worked with Realm in Swift projects in the past where I had to be aware not to pass objects or collections around and also to not use the same realm from multiple threads, etc. However, I don't think those issues apply to RN. Right?

My general approach is to have a Realm factory like:

**realm.js**

    export function getUserRealm(){
    
        const realm = new Realm({
            schema: combinedSchemas,
            path: `main.realm`
        });
    
        console.log( "💿 User Realm Path:", realm.path );
    
        return realm;
    
    }

Then I call `getUserRealm()` from wherever it is that I'm using realm.

There are a few places where I do pass `Results` between functions and callbacks.

*The main question is*, what should I guard against in order to prevent crashes?

**Comment by fealebenpae** [comment]
Hey @thatjuan,

I'm afraid that we cannot really guard from crashes like that in JavaScript. Native crashes are signs of a bug, not an expected error condition - those are signified by JavaScript errors that you can `try/catch`. Programmer error when using Realm shouldn't lead to native crashes.

And yes, threading is not an issue for Realm on React Native because there is only ever one thread that executes JavaScript.

We would be very grateful if you could share some more details about the crash, such as any stack trace that might be available when running your project from Xcode.

**Comment by thatjuan** [comment]
Thanks for pitching in @fealebenpae 

What I was referring to by guarding was not "catching" them, but rather preventing them.

I'd be happy to try and set up a constrained environment to reproduce the crashes, but for now, this is how I fixed them.

As I mentioned, I had set up a `getUserRealm()` function that returned a new Realm. I was calling that from every function that needed to use a realm. I did this because I accustomed to working with Realm on threaded environments, and this is generally what works best in those.

After coming back to the Realm-js docs and looking at the examples, I found that most of them were just creating one global Realm instance per file. I got rid of all the `getUserRealm()` calls and voilá.

I haven't had any crashes since doing this. However, I think this doesn't fix the bug. It just makes it less likely to surface. I say this because the crash wasn't always happening during the same step. It was randomly happening on reads or writes. Different tables, different triggers. I can only speculate, but it sounds like some sort of race condition or conflict caused by having more Realm instances.
